src/org/thoughtcrime/securesms/loki/protocol/ClosedGroupsProtocolV2.kt

```diff
@@ -87,7 +90,154 @@ object ClosedGroupsProtocolV2 {
+            job.setContext(context)
+            job.onRun() // Run the job immediately
+            // Send closed group update messages to any new members individually
+            for (member in membersToAdd) {
+                @Suppress("NAME_SHADOWING")
+                val closedGroupNewKind = ClosedGroupUpdateMessageSendJobV2.Kind.New(Hex.fromStringCondensed(groupPublicKey), name, encryptionKeyPair, membersAsData, adminsAsData)
+                @Suppress("NAME_SHADOWING")
+                val newMemberJob = ClosedGroupUpdateMessageSendJobV2(member, closedGroupNewKind)
+                ApplicationContext.getInstance(context).jobManager.add(newMemberJob)
+            }
+            // Notify the user
+            val infoType = GroupContext.Type.UPDATE
+            val threadID = DatabaseFactory.getThreadDatabase(context).getOrCreateThreadIdFor(Recipient.from(context, Address.fromSerialized(groupID), false))
+            insertOutgoingInfoMessage(context, groupID, infoType, name, updatedMembers, admins, threadID)
+        }
+    }
+
+    @JvmStatic
+    fun explicitRemoveMembers(context: Context, groupPublicKey: String, membersToRemove: List<String>): Promise<Any, Exception> {
+        return task {
+            val apiDB = DatabaseFactory.getLokiAPIDatabase(context)
+            val groupDB = DatabaseFactory.getGroupDatabase(context)
+            val groupID = doubleEncodeGroupID(groupPublicKey)
+            val group = groupDB.getGroup(groupID).orNull()
+            if (group == null) {
+                Log.d("Loki", "Can't leave nonexistent closed group.")
+                return@task Error.NoThread
+            }
+            val updatedMembers = group.members.map { it.serialize() }.toSet() - membersToRemove
+            groupDB.updateMembers(groupID, updatedMembers.map { Address.fromSerialized(it) })
+            val removeMembersAsData = membersToRemove.map { Hex.fromStringCondensed(it) }
+            val admins = group.admins.map { it.serialize() }
+            val encryptionKeyPair = apiDB.getLatestClosedGroupEncryptionKeyPair(groupPublicKey)
+            if (encryptionKeyPair == null) {
+                Log.d("Loki", "Couldn't get encryption key pair for closed group.")
+                return@task Error.NoKeyPair
+            }
+            if (membersToRemove.any { it in admins } && updatedMembers.isNotEmpty()) {
+                Log.d("Loki", "Can't remove admin from closed group unless the group is destroyed entirely.")
+                return@task Error.InvalidUpdate
+            }
+            val name = group.title
+            // Send the update to the group
+            val memberUpdateKind = ClosedGroupUpdateMessageSendJobV2.Kind.RemoveMembers(removeMembersAsData)
+            val job = ClosedGroupUpdateMessageSendJobV2(groupPublicKey, memberUpdateKind)
+            job.setContext(context)
+            job.onRun() // Run the job immediately
+            // Save the new group members
```

Comment:
If the current user is the group admin they should create and distribute a new key pair here